src/Frame.h

```diff
@@ -254,11 +253,8 @@ class Frame {
     tbb::queuing_rw_mutex _cache_mutex; // allow concurrent reads but lock for write
     std::mutex _image_mutex;            // only one disk access at a time
 
-    // Z profile counter, so Frame is not destroyed until finished
-    std::atomic<int> _z_profile_count;
-
-    // Moments counter, so Frame is not destroyed until finished
-    std::atomic<int> _moments_count;
+    // Use a shared lock for long time calculations, use an exclusive lock for the object destruction
+    mutable std::shared_mutex _life_mutex;
```

Comment:
I would suggest renaming "life mutex" to something like "calculation mutex" or "active task mutex".

Reply:
> I would prefer to rename the mutex to something that more clearly describes its function, but apart from that everything looks good.

agreed.